Add more constructors for diagTensor() and delta()

mtfishman opened the pull request:
(To create consistency with ITensor constructors.)

emstoudenmire commented:
Hi Matt,
One idea to improve this could be to just make the delta(IQIndexSet const& is); function signature be delta(IQIndexSet is); instead. Then, because IQIndexSet can be constructed from array, initializer_list, etc I think those cases may then automatically work without requiring extra code. The slight downside is that it will cause a copy for the case that the argument really is already just an IQIndexSet. But I think this will be a very slight extra cost, if any. What do you think?

mtfishman commented:
I don't have a strong opinion either way (and it is definitely a bit annoying defining many functions by hand so a more general approach would be nice). 

However, I just tried a test where I changed the signature to delta(IndexSet is), removed the version explicitly taking a vector, and ran the following:

```
auto i = Index("i",2);
auto j = Index("j",2);
std::vector<Index> inds = {i,j};
auto I = delta(inds);
```

and I get an error "no matching function for call to 'delta(std::vector<itensor::Index, std::allocator<itensor::Index> >&)'". This is followed by: "note:   no known conversion for argument 1 from 'std::vector<itensor::Index, std::allocator<itensor::Index> >' to 'itensor::IndexSet {aka itensor::IndexSetT<itensor::Index>}'". It is weird, because you can definitely call IndexSet(inds) to convert to an IndexSet (i.e. delta(IndexSet(inds)) works as expected). Maybe the conversion is being done in a way that C++ doesn't recognize it?

Anyway, I am inclined to just keep it as is, and if we find there are too many functions later on we can investigate a more generic solution based on conversions.

emstoudenmire commented:
I see - yes this is because that IndexSet constructor is marked explicit, so requires one to explicitly write IndexSet(v) where v is a std::vector. I could remove the explicit keyword, but maybe that will lead to too much "magic" i.e. unexpected conversions in other places. So I agree, let's go with the pull request as it is; it's already very reasonable, I was just trying to improve.
